### PR TITLE
fix(39332): QuoteStyle not respected for string literal types in implement interface

### LINKED ISSUE #39332: QuoteStyle not respected for string literal types in implement interface
From https://github.com/microsoft/vscode/issues/101219

**TypeScript Version:** 4.0.0-dev.20200629

**Search Terms:**

**Code**
Using `"typescript.preferences.quoteStyle": "single"`

```ts
interface IFoo {
  a(x: 'abc') : void;
}

class Foo implements IFoo {

}
```

Run implement interface on `Foo`

**Bug:**
The parameter value uses double quotes:

```ts
class Foo implements IFoo {
  a(x: "abc"): void {
    throw new Error('Method not implemented.');
  }
}
```

Also, if `typescript.preferences.quoteStyle` is unset, it would be nice if we used the original quote style for implement interface

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #39332

### CODE REVIEW

File: src/services/codefixes/helpers.ts [hunk 1]
```diff
@@ -143,14 +143,15 @@ namespace ts.codefix {
                 break;
         }
 
-        function outputMethod(signature: Signature, modifiers: NodeArray<Modifier> | undefined, name: PropertyName, body?: Block): void {
-            const method = signatureToMethodDeclaration(context, signature, enclosingDeclaration, modifiers, name, optional, body, importAdder);
+        function outputMethod(signature: Signature, modifiers: NodeArray<Modifier> | undefined, name: PropertyName, preferences: UserPreferences, body?: Block): void {
```
Comment (DanielRosenwasser) [inline src/services/codefixes/helpers.ts:146]: Nit: I think that `preferences` should be either the first or last parameter here

Comment (a-tarasyuk) [reply src/services/codefixes/helpers.ts:146]: Oke, I think it should be first because there is the last optional `body` parameter. Is it ok?

Comment (DanielRosenwasser) [reply src/services/codefixes/helpers.ts:146]: Sure!

**Comment by minestarks** [comment]
@jessetrinity  could you review please?

**Comment by a-tarasyuk** [comment]
@jessetrinity Good question. Which result should be generated for the following code with the `auto` option?

```ts
interface I {
   a(): void;
   b(x: 'x', y: 'a' | 'b'): 'b';

   c: 'c';
   d: { e: 'e'; };
}

class Foo/**/ implements I {
}
```

**Comment by jessetrinity** [comment]
@a-tarasyuk If auto were respected I would expect the output to be based on the first use of quotes in an interface member. In this example, single quotes since you have `b(x: 'x', y: 'a' | 'b'): 'b';`, or double quotes if it were instead `b(x: "x", y: "a" | "b"): "b";`.

I'm not sure if that decision makes sense though since I think `auto` might only apply to imports?

**Comment by a-tarasyuk** [comment]
@jessetrinity Do you mean we need to find the use of the first line in the file to determine the `quotePreference` before generating the code?

**Comment by jessetrinity** [comment]
@a-tarasyuk that is the usual way of doing it. I think `getQuotePreference` in `utilities` is what you want.

**Comment by a-tarasyuk** [comment]
@jessetrinity I checked this helper, it seems to detect `quotePreference` based on` imports`. If it is ok, I'll use it to get quote option.

**Comment by jessetrinity** [comment]
@a-tarasyuk That seems okay. It looks like the setting is worded to only work based on imports in editors anyway.

**Comment by a-tarasyuk** [comment]
@jessetrinity Ok, changed and added several additional tests for `auto` option.

**Comment by a-tarasyuk** [comment]
@jessetrinity Ok, changed and added several additional tests for auto option.

File: tests/cases/fourslash/fourslash.ts [hunk 2]
```diff
@@ -592,7 +592,7 @@ declare namespace FourSlashInterface {
         filesToSearch?: ReadonlyArray<string>;
     }
     interface UserPreferences {
-        readonly quotePreference?: "double" | "single";
+        readonly quotePreference?: "auto" | "double" | "single"
```
Comment (jessetrinity) [inline tests/cases/fourslash/fourslash.ts:595]:
```suggestion
        readonly quotePreference?: "auto" | "double" | "single";

```

**Comment by jessetrinity** [comment]
I think this commit will fix the eslint failures:
https://github.com/microsoft/TypeScript/pull/39450

**Pull request merged**